### PR TITLE
Epic/mtp refactor phase1

### PR DESCRIPTION
**Pull request by pengchengneo**
## Epic: MTP Refactor Phase 1

Multi-layer speculative decoding (MTP/NEXTN) for MiMo-V2 models on TPU.

### Commits

1. **`b62b465f` — MultiLayerEAGLEWorker / MultiLayerDraftWorker (#1089)**
   N draft model runners (one per mtp_layer_idx) with 3-layer MTP support for MiMo-V2.5-Pro.

2. **`20287f5f` — NEXTN E2E test (#1097)**
   Environment-gated E2E test for V2.5-Pro NEXTN on v6e-64 multi-host.

3. **`f5d40219` — DP-aware spec decode (#1108)**
   DP-correct padding, per-rank spec_info split/concat, and ScheduleBatch integration for dp_size > 1.

4. **`6e854218` — Per-rank ScheduleBatch.spec_info (#1171)**
   Refactor spec_info into per-rank split/concat helpers (option C), eliminating cross-rank-flat layout.

5. **`6e15718a` — MiMo V2 Flash MTP loading (#1187)**
   Fix draft model weight loading for MiMo-V2-Flash (hybrid SWA architecture).

6. **`9cc3a163` — Fix all cache miss and precompile shapes (#1188)**
   Eliminate post-precompile JIT cache misses: accept_lens pytree alignment, capture_hidden_mode consistency, spec_headroom before dp-scale, draft_runner_cache_size before dp-scale, and pad_to_bucket for spec decode.

### Eval (GSM8K via evalscope)

| Setup | Model | dp | spec | GSM8K Accuracy |
|-------|-------|----|------|----------------|
| Single-host v6e-4 | Llama-3.1-8B-Instruct + EAGLE3 | 1 | EAGLE | 0.81 |
| Single-host v6e-4 | Llama-3.1-8B-Instruct | 1 | nospec | 0.80 |
| Single-host v6e-4 | Llama-3.1-8B-Instruct + EAGLE3 | 2 | EAGLE | 0.81 |
| Single-host v6e-4 | Llama-3.1-8B-Instruct | 2 | nospec | 0.82 |
| Multi-host v6e-256 | MiMo-V2-Flash + NextN | 1 | NEXTN | 0.98 |
| Multi-host v6e-256 | MiMo-V2-Flash | 1 | nospec | 0.98 |
| Multi-host v6e-256 | MiMo-V2-Flash + NextN | 2 | NEXTN | 0.98 |
| Multi-host v6e-256 | MiMo-V2-Flash | 2 | nospec | 0.98 |

### Phase 2 (follow-up)

To keep this PR reviewable, the following work is deferred to Phase 2:

- #1194 — `build_eagle_tree_structure` FULL_MASK OOM at large batch sizes
- #1113 — Eagle3 topk > 1 support
- #1109 — Eagle3 non-greedy sampling

### CODE REVIEW

**Review by zorrofox — COMMENTED** [review]
The 5 commits are all individually reviewed (#1089/#1097/#1108 ours, #1171 reviewed + hw-verified, #1187 just looked at — the fused-vs-split QKV detection for V2-Flash MTP is straightforward and has its own test). Old CI run 26273411747 is fully green (unit/e2e/perf/accuracy).

Two things before merge:

1. **Needs rebase on main** — `coordinator / decide` fails with `can't open file scripts/ci/coordinator_decide.py: No such file or directory`. That script landed on main via #1156 (5/20) and epic is 14 commits behind. Mergeable says no conflicts so should be a clean rebase.

2. **Wait for #1188** (per your plan) — though note #1188 has the topk>1 tree-mask issue I just flagged there. If you want to decouple, this PR as-is (5 commits, no #1188) is already a coherent Phase-1 milestone: P1-0/1/3/5a were merged into main directly back in 5/14 (#1063-66), and these 5 are P1-4/5b/8 + the per-rank refactor + V2-Flash MTP loading. Landing this first and doing #1188 as a follow-up to main would unblock #1113 (topk>1) from rebasing onto a stable base.

Nit: PR body is empty — a one-paragraph "Phase-1 summary + what's deferred to Phase-2 (#1109/#1113/#1090)" would help future archaeology.

File: python/sgl_jax/srt/speculative/eagle_util.py [hunk 1]
```diff
@@ -635,16 +670,19 @@ def prepare_extend_after_decode(
             return
 
         batch.input_ids = self.verified_id
-        accept_length_cpu_arr = batch.spec_info.accept_length_cpu
+        rank0_spec = batch.reqs_info[0].spec_info if batch.reqs_info else None
```
[the remark this answers is not in the document]
Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/eagle_util.py]: removed

File: python/sgl_jax/srt/configs/model_config.py [hunk 2]
```diff
@@ -142,8 +142,21 @@ def __init__(
...
+            # Each draft runner is a single SWA layer; without this the KV pool
+            # sizes for all 70 target layers and OOMs.
+            self.hf_config.num_hidden_layers = 1
+            if self.quantization_config is not None:
+                # eh_proj / o_proj are BF16 in model_mtp.safetensors (no
+                # weight_scale_inv); keep them as plain LinearBase so mappings
+                # can target `.weight` instead of `.weight_q`.
+                ignored = list(self.quantization_config.ignored_layers or [])
+                ignored.extend(["model.eh_proj", "model.mtp_block.self_attn.o_proj"])
+                self.quantization_config.ignored_layers = ignored
```
[the remark this answers is not in the document]
Comment (pengchengneo) [reply python/sgl_jax/srt/configs/model_config.py:159]: currently this is safe

File: python/sgl_jax/srt/speculative/multi_layer_draft_worker.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+            cur_hidden = logits_output.hidden_states
+
+        # logits_indices_selector maps global-flat req k → DP-padded slot s_k.
+        # rep_logits/rep_hidden are DP-padded (total_bs, …)/(total_bs*(steps+1), …);
+        # keep topk_probs_from_logits running on device with the padded shape
+        # (varying real_bs gather on device would generate a fresh trace per
+        # warmup batch size — see #1090). Gather to real_bs on host.
+        sel = np.asarray(model_worker_batch.logits_indices_selector)
+        accept_host = np.asarray(jax.device_get(batch_output.accept_lens))
+        select_index = sel * (self.speculative_num_steps + 1) + accept_host[sel] - 1
```
[the remark this answers is not in the document]
Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/multi_layer_draft_worker.py:328]: added assert

File: python/sgl_jax/srt/speculative/multi_layer_draft_worker.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+                i, score_list, token_list, parents_list, tree_info, self.topk
+            )
+            if i == self.speculative_num_steps - 1:
+                break
+
+            forward_batch = update_forward_batch_info(
+                forward_batch, i, input_ids, hidden_states, positions_base
+            )
+            mr = self.runner(i)
+            mr.attn_backend.forward_metadata = metadata_per_layer[i][i]
```
Comment (cjx0709) [inline python/sgl_jax/srt/speculative/multi_layer_draft_worker.py:180]: should forward_batch.attention_backend also should updated to mr.attn_backend? otherwise layer1/layer2 will use layer0.attention_backend to compute

File: python/sgl_jax/srt/managers/schedule_batch.py [hunk 5]
```diff
@@ -2046,6 +2047,237 @@ def _merge_sampling_info(
...
+            token_ids_logprobs=None,
+            sampling_info=sampling_info,
+            positions=np.empty(0, dtype=np.int32),
+            cache_loc=np.empty(0, dtype=np.int32),
+            extend_prefix_lens=None,
+            extend_seq_lens=None,
+            extend_logprob_start_lens=None,
+            extend_input_logprob_token_ids=None,
+            logits_indices=None,
+            lora_ids=(
```
Comment (cjx0709) [inline python/sgl_jax/srt/managers/schedule_batch.py:2149]: `lora_ids` here seems to stay in flat real-request order, but this batch is otherwise already in DP-padded slot order. For unbalanced DP batches, does that misalign LoRA adapter selection with `seq_lens` / slot index?

E.g. `bs_per_rank=[1,2]` gives slots `[A, PAD, B, C]`, but this would produce `['A','B','C','0']`.

Comment (pengchengneo) [reply python/sgl_jax/srt/managers/schedule_batch.py:2149]: currently, spec_decode is not compatible with lora, will fix it future

File: python/sgl_jax/srt/speculative/multi_layer_draft_worker.py (hunk 4 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/multi_layer_draft_worker.py:180]: yes, fixd it in 37a7245f

File: python/sgl_jax/srt/speculative/multi_layer_draft_worker.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+                layer0_out = logits_output
+
+        # Next-round draft state = layer 0's last-token hidden + topk (draft step 0
+        # starts from layer 0 with target's last hidden, so we cache layer 0's
+        # output here so step 0 can be skipped).
+        rep_logits, rep_hidden = replicate_to_mesh(
+            self.mesh, layer0_out.next_token_logits, layer0_out.hidden_states
+        )
+        layer0_out.next_token_logits = rep_logits
+        layer0_out.hidden_states = rep_hidden[last_idx]
```
Comment (cjx0709) [inline python/sgl_jax/srt/speculative/multi_layer_draft_worker.py:269]: the `last_idx` is still rank-local here, but `rep_hidden` is already global across DP ranks. For `dp>1`, rank>0 may miss its token offset and read another rank’s last hidden state.

File: python/sgl_jax/srt/speculative/eagle_draft_worker.py [hunk 7]
```diff
@@ -449,11 +472,9 @@ def draft_forward(self, model_worker_batch: ModelWorkerBatch):
         return score_list, token_list, parents_list
 
     def _pick_context_len(self, max_seq_len: int) -> int:
-        max_seq_len = max(int(max_seq_len), 1)
         if self.precompile_token_paddings:
-            for padding in self.precompile_token_paddings:
-                if padding >= max_seq_len:
-                    return padding
+            return self.precompile_token_paddings[-1]
```
Comment (cjx0709) [inline python/sgl_jax/srt/speculative/eagle_draft_worker.py:476]: `precompile_token_paddings[-1]` is a total-token bucket, not a per-request context bound. Using it as `max_context_len` here seems to blow up FULL_MASK capacity by another factor of `bs`, may cause OOM. Should this still bucket off runtime `max_seq_len` instead?

Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/eagle_draft_worker.py:476]: precompile_token_paddings[-1] here is intentional — it fixes tree_mask_capacity to a static shape so the Pallas kernel compilation can be reused across batches. If we bucket off runtime max_seq_len, every distinct value triggers a retrace/recompile, which is more expensive than the extra memory allocation on TPU. The capacity is allocated once and reused, so no OOM risk in practice.


In production environment on various accelerators,  the tree-attention is not used widely, in phase2 we will optim top=1 path to avoid this problem, currently I think this is ok

File: python/sgl_jax/srt/speculative/multi_layer_draft_worker.py (hunk 6 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/multi_layer_draft_worker.py:269]: fix it e275871c

**Review by cjx0709 — APPROVED** [review]
LGTM
